### DISCUSSION TITLE
Comparison between supported WSL blends

### DISCUSSION THREAD

**Discussion by ssbarnea** [Vampire/setup-wsl#39] [General]
I observed that the default WSL distro setup-wsl uses is `Debian`, without mentioning which version it would be.

On the other hand, when using WSL under Windows, they seem to default on Ubuntu (while still not mentioning which version).

I am quite curious about pros/cons between the deb based variants like: Debian, ubuntu-22.04 and ubuntu-20.04. One interesting aspect would likely be, which one will attempt to download less updates when doing `apt-get update`.

![](https://sbarnea.com/ss/Screen-Shot-2022-07-04-14-07-57.19.png)

Did we have a special reason for going for a different default than Windows itself?

**Comment by Vampire** [comment]
> I observed that the default WSL distro setup-wsl uses is `Debian`, without mentioning which version it would be.

The Microsoft store doesn't mention a version either: https://apps.microsoft.com/store/detail/debian/9MSVKQC78PK6
So I followed suit, assuming that the version might maybe change over time.
For Ubuntu there is one package per LTS version.

> On the other hand, when using WSL under Windows, they seem to default on Ubuntu (while still not mentioning which version).

Exactly

> I am quite curious about pros/cons between the deb based variants like: Debian, ubuntu-22.04 and ubuntu-20.04.

Probably not too much differences for most cases, unless you need something specific to a distro or just want to matrix-test on all to make sure it works.

> One interesting aspect would likely be, which one will attempt to download less updates when doing `apt-get update`.

If you really mean `update`, probably not much difference, as it only downloads the current list of packages and versions.
If you mean `upgrade`, then it heavily depends and changes over time. There are constantly new versions added until a version is end of life. But as I said I'm also not sure whether maybe the installed OS version will change over time which would also change it again. 
But generally said, Debian is more strict in what it adds for already released versions. They usually only add security patches and not even new feature-versions, as a Debian package release strategy and distro - other than Ubuntu's - is more focused on stability than to have always the latest versions.

> Did we have a special reason for going for a different default than Windows itself?

Probably mainly personal preference. On desktop systems I prefer Ubuntu as it is more end-user friendly, and more permissive with package updates. On servers and automations I prefer Debian, as it is imho more suited for that use-case and more designed for stability.

Ubuntu is afaik mainly historically the default, or maybe by agreement. Canonical, the creators of Ubuntu, afaik partnered with Microsoft to get the whole thing going and thus also was the first available WSL distribution.

**Reply by ssbarnea** [reply]
Thanks for clarifying these. I did some tests and as of today (2022-07-04):
* `Debian` version is 9, and if you install python3 you get `python3.5` ... LOL, already dead
* `Ubuntu` version is 20.04, so you get `python3.8` on it, which is ok

That makes be think that Microsoft did the right choice and pick Ubuntu instead of Debian. I will update the config to enforce use of Ubuntu on my wsl-setup calls, seems better for my own use case, where I rely on python.

BTW, I do not know how you managed to get `Ubuntu-22.04` because on my testing Windows 10 Pro with all updates, `wsl --list --online` does not show it as available. Even manually supplying it it does not work, so `22.04` is clearly not available for everyone.

**Reply by Vampire** [reply]
> * `Debian` version is 9, and if you install python3 you get `python3.5` ... LOL, already dead

Well, even Debian 9 is dead already and as I said, Debian is restrained with package updates within one Debian version where only security patches are applied. I guess Python minor version is feature release, not security patch, so Debian does not provide it as update within one version due to stability considerations.

But the Debian 9 can also be updated to a newer version of Debian of course using `dist-upgrade` and then it also provides newer versions of packages.

> That makes be think that Microsoft did the right choice and pick Ubuntu instead of Debian.

As I said, there are different considerations and pros and cons for both. :-)

> I will update the config to enforce use of Ubuntu on my wsl-setup calls, seems better for my own use case, where I rely on python.

Quite possible, that's why it is configurable. :-)

> BTW, I do not know how you managed to get `Ubuntu-22.04` because on my testing Windows 10 Pro with all updates, `wsl --list --online` does not show it as available. Even manually supplying it it does not work, so `22.04` is clearly not available for everyone.

Maybe that's a bug or cache issue or whatever. Also I do not use that functionality in the action. At https://apps.microsoft.com/store/detail/ubuntu-2204-lts/9PN20MSR04DW you see the store page of it. And at https://docs.microsoft.com/en-us/windows/wsl/install-manual#downloading-distributions you find a list of static short-links to some available distros that point to the actual download file that you can use to install the distro without integration or windows store which is what I do in the action too.

You can even install distros that do not have an explicit WSL variant with some extra effort.
